llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+			return &callEquivalentQueue[indexCallEquivalentQueue++];
+	}	
+	void visitCallEquivalent(const VectorOfArgs& arguments)
+	{
+		//Insert the arguments in the map
+		bool areAllArgsNullptr = true;
+		{
+			moduleData.setUpPartialInterpreter(F, currentEE);
+
+
+			assert(arguments.size() == F.getFunctionType()->getNumParams());
+
+			for (uint32_t i=0; i<arguments.size(); i++)
+			{
+				const llvm::Value* x = arguments[i];
+				if (!x)
+					continue;
+				areAllArgsNullptr = false;
+				llvm::Argument* ith_arg = F.getArg(i);
+				currentEE->assignToMaps(ith_arg, x);
+			}
+		//Do the visit
+		actualVisit(/*skipVisit*/areAllArgsNullptr);
+		
+		//Cleanup
+		doneVisitCallBase();
+		}
+
+	}
+	void actualVisit(bool skipVisit);
+	void visitCallBase(const llvm::CallBase* callBase)
+	{
+		const auto& equivalentArgs = getArguments(callBase, nullptr, nullptr);
+
+		enqueCallEquivalent(equivalentArgs);
+  		// Handle varargs arguments...
+	}
+	void enqueVisitNoInfo()
+	{
+		const auto& equivalentArgs = getArguments(nullptr, nullptr, nullptr);
+
+		enqueCallEquivalent(equivalentArgs);
+	}
+	void doneVisitCallBase()
+	{
+		bool removed = currentEE->removeModule(moduleData.getModulePtr());
+		(void)removed;
+		assert(removed);
+		delete currentEE;
+		currentEE = nullptr;
+	}
+	void cleanupBB(Function& function)
+	{
+		int BBs =0;
+		std::set<std::pair<llvm::BasicBlock*, llvm::BasicBlock*>> Edges;
+		for (auto& x : function)
+		{
+			BBs++;
+			for (auto* p : predecessors(&x))
+				Edges.insert({p, &x});
+		}
+
+		for (auto& p : Edges)
+
+		{
+			if (visitedEdges.count(p) == 0)
+				removeEdgeBetweenBlocks(p.first, p.second);
+		}
+
+	}
+	bool hasModifications(const bool emitStats) const
+	{
+		uint32_t numberBBs =0;
+		llvm::DenseSet<std::pair<const llvm::BasicBlock*, const llvm::BasicBlock*>> Edges;
+		for (const BasicBlock& x : F)
+		{
+			numberBBs++;
+			for (const BasicBlock* p : predecessors(&x))
+				Edges.insert({&x, p});
+		}
+
+		llvm::DenseSet<const llvm::BasicBlock*> visitedBBs;
+		visitedBBs.insert(&F.getEntryBlock());
+		for (const auto& pair : visitedEdges)
+			visitedBBs.insert(pair.second);
+
+		const uint32_t numberVisitedBBs = visitedBBs.size();
+		const uint32_t numberVisitedEdges = visitedEdges.size();
+
+		if ((numberVisitedBBs != numberBBs) || (Edges.size() != numberVisitedEdges))
+		{
+			if (emitStats)
+			{
+				llvm::errs() << "Total edges \t\t" << numberVisitedEdges << "\t" << Edges.size() <<"\n";
+				llvm::errs() << "Reachable BB\t\t" << numberVisitedBBs << "\t" << numberBBs << "\n";
+				llvm::errs() << "In function:\t\t" << F.getName()<< "\n";
+			}
+			return true;
+		}
+		return false;
+	}
+};
+
+void ModuleData::initFunctionData()
+{
+	assert(functionData.empty());
+	for (const Function& F : module)
+	{
+		functionData.emplace(&F, FunctionData(F, *this));
+	}
+}
+
+FunctionData& ModuleData::getFunctionData(const llvm::Function& F)
+{
+	return functionData.at(&F);
+}
+
+}//cheerp
+namespace llvm{
+
+
+class SubGraph;
```

Comment:
Do we actually need all this complexity?

Reply:
Added comment explaining the purpose of this (that we need to simplify SCCs)